enterprise/ha/src/test/java/org/neo4j/kernel/impl/ha/ClusterManager.java

```diff
@@ -309,17 +303,47 @@ private static int findFreePort( final int minPort, final int maxPort, final Set
                 // This port is already taken but not bound yet, ignore it
                 continue;
             }
+
+            try
+            {
+                // try binding it at wildcard
+                ServerSocket ss = new ServerSocket();
+                ss.setReuseAddress( false );
+                ss.bind( new InetSocketAddress( port ) );
+                ss.close();
+            }
+            catch ( IOException e )
+            {
+                except.add( port );
+                continue;
+            }
+
+            try
+            {
+                // try binding it at loopback
+                ServerSocket ss = new ServerSocket();
+                ss.setReuseAddress( false );
+                ss.bind( new InetSocketAddress( InetAddress.getLoopbackAddress(), port ) );
+                ss.close();
+            }
+            catch ( IOException e )
+            {
+                except.add( port );
+                continue;
+            }
+
             try
             {
-                ServerSocket socket = new ServerSocket( port );
-                // Port is available, return it
+                // try connecting to it at loopback
+                Socket socket = new Socket( InetAddress.getLoopbackAddress(), port );
                 socket.close();
-                return port;
+                except.add( port );
+                continue;
             }
             catch ( IOException ex )
             {
-                // Port was already bound, try the next one
-                except.add( port );
+                // Port very likely free!
```

Comment:
@mneedham This part of the code tries to open a connection, but if it fails to open a connection then it throws an exception, which is the desired outcome because it implies nobody is listening and accepting connections.